File: projects/swagger-studio-ruleset/python/src/swagger_studio_ruleset_publisher/packager.py
"""Bundle a ruleset directory for upload.

The on-disk layout is modular: `spectral.yaml` extends individual
`rules/*.yaml` category files for clean editing and review. Before upload
we flatten everything into a single self-contained `spectral.yaml` so
Studio doesn't have to resolve relative `./rules/*.yaml` references (it
may not, and the modular form is for humans, not the engine).

Built-in extends like `spectral:oas` are left in place — those refer to
Spectral's own bundled rulesets and resolve at runtime regardless.
"""

from __future__ import annotations

import shutil
import tempfile
from dataclasses import dataclass
from pathlib import Path
from typing import Any
from zipfile import ZIP_DEFLATED, ZipFile

import yaml


@dataclass(frozen=True)
class RulesetBundle:
    """Result of packaging: a validated directory + a zip artifact path."""

    directory: Path
    zip_path: Path


def validate(directory: Path) -> Path:
    """Resolve and sanity-check the ruleset directory.

    Returns the absolute path. Raises with a helpful message if the directory
    or required entry point is missing — these checks fail fast before any
    network call so the user gets a useful error.
    """
    resolved = directory.expanduser().resolve()
    if not resolved.is_dir():
        raise FileNotFoundError(f"Ruleset directory not found: {resolved}")
    entry = resolved / "spectral.yaml"
    if not entry.is_file():
        raise FileNotFoundError(
            f"Ruleset entry point not found: {entry}\n"
            "Every ruleset must declare a top-level spectral.yaml."
        )
    return resolved


def package(directory: Path, zip_dest: Path) -> RulesetBundle:
    """Validate, flatten extends, and produce a single-file ZIP at `zip_dest`.

    The ZIP contains exactly one file (`spectral.yaml`) with every relative
    extends inlined. No `rules/` subdirectory in the artifact — modular form
    is for humans, flattened form is for Studio.
    """
    resolved = validate(directory)
    merged_yaml = flatten(resolved)

    zip_dest.parent.mkdir(parents=True, exist_ok=True)
    if zip_dest.exists():
        zip_dest.unlink()

    with ZipFile(zip_dest, "w", ZIP_DEFLATED) as zf:
        zf.writestr("spectral.yaml", merged_yaml)

    return RulesetBundle(directory=resolved, zip_path=zip_dest)


def flatten(directory: Path) -> str:
    """Resolve all relative `extends: ./...yaml` references inline.

    Returns the merged YAML content as a string. Built-in extends (e.g.
    `spectral:oas`, `[spectral:oas, recommended]`) are kept as-is.

    Rule conflicts (same name from multiple files) take last-write-wins
    in extends order, with the entry file's own `rules:` block applied
    last so it always overrides.
    """
    main = yaml.safe_load((directory / "spectral.yaml").read_text(encoding="utf-8")) or {}
    merged_rules: dict[str, Any] = {}
    runtime_extends: list[Any] = []

    raw_extends = main.get("extends") or []
    if not isinstance(raw_extends, list):
        raw_extends = [raw_extends]

    for entry in raw_extends:
        if _is_relative_file_ref(entry):
            sub_path = (directory / entry).resolve()
            sub = yaml.safe_load(sub_path.read_text(encoding="utf-8")) or {}
            sub_rules = sub.get("rules") or {}
            if isinstance(sub_rules, dict):
                merged_rules.update(sub_rules)
        else:
            # built-in ruleset reference — leave for Spectral to resolve.
            runtime_extends.append(entry)

    # Inline rules from the main file last, so they win conflicts.
    own_rules = main.get("rules") or {}
    if isinstance(own_rules, dict):
        merged_rules.update(own_rules)

    output: dict[str, Any] = {}
    if runtime_extends:
        output["extends"] = runtime_extends
    if merged_rules:
        output["rules"] = merged_rules
    return yaml.safe_dump(output, sort_keys=False, default_flow_style=False)


def write_flattened_dir(directory: Path) -> Path:
    """Materialize the flattened ruleset to a temp directory.

    Returns the path to the temp directory (containing a single
    `spectral.yaml`). Caller is responsible for cleanup — use as a
    `shutil.rmtree` target or pass to the CLI backend.
    """
    merged_yaml = flatten(directory)
    tmp = Path(tempfile.mkdtemp(prefix="ruleset-flat-"))
    (tmp / "spectral.yaml").write_text(merged_yaml, encoding="utf-8")
    return tmp


def _is_relative_file_ref(entry: Any) -> bool:
    """True iff this extends entry points at a local YAML file."""
    if not isinstance(entry, str):
        return False
    return entry.startswith("./") or entry.startswith("../") or entry.endswith((".yaml", ".yml"))


def cleanup(bundle: RulesetBundle) -> None:
    """Best-effort cleanup of the generated zip. Safe to call multiple times."""
    try:
        bundle.zip_path.unlink(missing_ok=True)
    except OSError:
        pass


def temp_zip_path(directory: Path) -> Path:
    """Choose a deterministic but isolated location for the temp zip."""
    return directory.parent / f".{directory.name}.bundle.zip"


def has_swaggerhub_cli() -> bool:
    """Cheap availability check for the CLI backend."""
    return shutil.which("swaggerhub") is not None

File: projects/swagger-studio-ruleset/typescript/src/packager.ts
/**
 * Bundle a ruleset directory for upload.
 *
 * The on-disk layout is modular: `spectral.yaml` extends individual
 * `rules/*.yaml` category files for clean editing and review. Before upload
 * we flatten everything into a single self-contained `spectral.yaml` so
 * Studio doesn't have to resolve relative `./rules/*.yaml` references.
 *
 * Built-in extends like `spectral:oas` are left in place — those refer to
 * Spectral's own bundled rulesets and resolve at runtime regardless.
 */

import archiver from "archiver";
import { createWriteStream, existsSync, mkdtempSync, statSync, unlinkSync, writeFileSync } from "node:fs";
import { access, readFile } from "node:fs/promises";
import { tmpdir } from "node:os";
import { dirname, join, resolve } from "node:path";
import * as YAML from "yaml";

export interface RulesetBundle {
  readonly directory: string;
  readonly zipPath: string;
}

export async function validate(directory: string): Promise<string> {
  const resolved = resolve(directory);
  if (!existsSync(resolved) || !statSync(resolved).isDirectory()) {
    throw new Error(`Ruleset directory not found: ${resolved}`);
  }
  const entry = join(resolved, "spectral.yaml");
  if (!existsSync(entry)) {
    throw new Error(
      `Ruleset entry point not found: ${entry}\n` +
        "Every ruleset must declare a top-level spectral.yaml.",
    );
  }
  return resolved;
}

/**
 * Resolve all relative `extends: ./...yaml` references inline and return the
 * merged YAML content. Built-in extends (e.g. `spectral:oas`) are kept.
 * Own `rules:` in the entry file always wins conflicts.
 */
export async function flatten(directory: string): Promise<string> {
  const resolved = await validate(directory);
  const main = (YAML.parse(await readFile(join(resolved, "spectral.yaml"), "utf-8")) ?? {}) as Record<
    string,
    unknown
  >;

  const mergedRules: Record<string, unknown> = {};
  const runtimeExtends: unknown[] = [];

  const rawExtends = normalizeArray(main["extends"]);
  for (const entry of rawExtends) {
    if (isRelativeFileRef(entry)) {
      const subPath = resolve(resolved, entry);
      const sub = (YAML.parse(await readFile(subPath, "utf-8")) ?? {}) as Record<string, unknown>;
      const subRules = sub["rules"];
      if (isRecord(subRules)) {
        Object.assign(mergedRules, subRules);
      }
    } else {
      runtimeExtends.push(entry);
    }
  }

  const ownRules = main["rules"];
  if (isRecord(ownRules)) {
    Object.assign(mergedRules, ownRules);
  }

  const output: Record<string, unknown> = {};
  if (runtimeExtends.length > 0) {
    output["extends"] = runtimeExtends;
  }
  if (Object.keys(mergedRules).length > 0) {
    output["rules"] = mergedRules;
  }
  return YAML.stringify(output);
}

/**
 * Materialize the flattened ruleset to a temp directory containing a single
 * `spectral.yaml`. Caller is responsible for cleanup.
 */
export async function writeFlattenedDir(directory: string): Promise<string> {
  const merged = await flatten(directory);
  const tmp = mkdtempSync(join(tmpdir(), "ruleset-flat-"));
  writeFileSync(join(tmp, "spectral.yaml"), merged, "utf-8");
  return tmp;
}

/**
 * Validate, flatten, and write a single-file ZIP at `zipDest`. The ZIP
 * contains exactly one file (`spectral.yaml`) with rules inlined.
 */
export async function pkg(directory: string, zipDest: string): Promise<RulesetBundle> {
  const resolved = await validate(directory);
  const merged = await flatten(resolved);

  if (existsSync(zipDest)) {
    unlinkSync(zipDest);
  } else {
    await access(dirname(zipDest)).catch(async () => {
      const { mkdir } = await import("node:fs/promises");
      await mkdir(dirname(zipDest), { recursive: true });
    });
  }

  await new Promise<void>((res, rej) => {
    const output = createWriteStream(zipDest);
    const archive = archiver("zip", { zlib: { level: 9 } });
    output.on("close", () => {
      res();
    });
    archive.on("error", (err) => {
      rej(err);
    });
    archive.pipe(output);
    archive.append(merged, { name: "spectral.yaml" });
    void archive.finalize();
  });

  return { directory: resolved, zipPath: resolve(zipDest) };
}

export function cleanup(bundle: RulesetBundle): void {
  try {
    if (existsSync(bundle.zipPath)) {
      unlinkSync(bundle.zipPath);
    }
  } catch {
    // best-effort cleanup
  }
}

export function tempZipPath(directory: string): string {
  const resolved = resolve(directory);
  const parent = dirname(resolved);
  const name = resolved.split("/").pop() ?? "ruleset";
  return join(parent, `.${name}.bundle.zip`);
}

export async function hasSwaggerhubCli(): Promise<boolean> {
  const { execFile } = await import("node:child_process");
  return new Promise<boolean>((res) => {
    execFile("swaggerhub", ["--version"], (err) => {
      res(err === null);
    });
  });
}

// --- helpers -----------------------------------------------------------

function normalizeArray(value: unknown): unknown[] {
  if (Array.isArray(value)) return value;
  if (value == null) return [];
  return [value];
}

function isRelativeFileRef(entry: unknown): entry is string {
  if (typeof entry !== "string") return false;
  return entry.startsWith("./") || entry.startsWith("../") || entry.endsWith(".yaml") || entry.endsWith(".yml");
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}
